datafusion-examples/examples/struct_field_rewrite.rs

```diff
@@ -0,0 +1,353 @@
+    }
+}
+
+/// Rewriter that converts struct field access to flattened column references
+#[derive(Debug)]
+struct StructFieldRewriter;
+
+impl FileExpressionRewriter for StructFieldRewriter {
+    fn rewrite(
+        &self,
+        _table_schema: SchemaRef,
+        file_schema: SchemaRef,
+        expr: Arc<dyn PhysicalExpr>,
+    ) -> Result<Arc<dyn PhysicalExpr>> {
+        let mut rewrite = StructFieldRewriterImpl { file_schema };
+        expr.rewrite(&mut rewrite).data()
+    }
+}
+
+struct StructFieldRewriterImpl {
+    file_schema: SchemaRef,
+}
+
+impl TreeNodeRewriter for StructFieldRewriterImpl {
+    type Node = Arc<dyn PhysicalExpr>;
+
+    fn f_down(
+        &mut self,
+        expr: Arc<dyn PhysicalExpr>,
+    ) -> Result<Transformed<Arc<dyn PhysicalExpr>>> {
+        if let Some(scalar_function) = expr.as_any().downcast_ref::<ScalarFunctionExpr>()
+        {
+            if scalar_function.name() == "get_field" {
+                if scalar_function.args().len() == 2 {
+                    // First argument is the column, second argument is the field name
+                    let column = scalar_function.args()[0].clone();
+                    let field_name = scalar_function.args()[1].clone();
+                    if let Some(literal) =
+                        field_name.as_any().downcast_ref::<expressions::Literal>()
+                    {
+                        if let Some(field_name) = literal.value().try_as_str().flatten() {
+                            if let Some(column) =
+                                column.as_any().downcast_ref::<expressions::Column>()
+                            {
+                                let column_name = column.name();
+                                let source_field =
+                                    self.file_schema.field_with_name(column_name)?;
+                                let expected_flattened_column_name =
+                                    format!("_{}.{}", column_name, field_name);
+                                // Check if the flattened column exists in the file schema and has the same type
+                                if let Ok(shredded_field) = self
+                                    .file_schema
+                                    .field_with_name(&expected_flattened_column_name)
+                                {
+                                    if source_field.data_type()
+                                        == shredded_field.data_type()
+                                    {
+                                        // Rewrite the expression to use the flattened column
+                                        let rewritten_expr = expressions::col(
+                                            &expected_flattened_column_name,
+                                            &self.file_schema,
+                                        )?;
+                                        return Ok(Transformed::yes(rewritten_expr));
```

Comment:
This is completely arbitrary: this could return another expression, it doesn't _have_ to be just a column.